src/Npgsql/NpgsqlTypes/NpgsqlCube.cs

```diff
@@ -0,0 +1,155 @@
+        /// </summary>
+        public IEnumerable<double> LowerLeft => _coords.Take(Dimensions);
+        
+        /// <summary>
+        /// The upper right coordinates of the cube.
+        /// </summary>
+        public IEnumerable<double> UpperRight => _coords.Skip(Point ? 0 : Dimensions).Take(Dimensions);
+
+        /// <summary>
+        /// Checks if the cube is a point.
+        /// </summary>
+        /// <returns></returns>
+        internal bool IsPointInternal()
+        {
+            if (Point) return true;
+            
+            /*
+             * Even if the point-flag is not set, all the lower-left coordinates might
+             * match the upper-right coordinates, so that the value is in fact a
+             * point. Such values don't arise with current code - the point flag is
+             * always set if appropriate - but they might be present on-disk in
+             * clusters upgraded from pre-9.4 versions.
+             */
+            for (var i = 0; i < Dimensions; i++)
+            {
+                if (LowerLeft.ElementAt(i) != UpperRight.ElementAt(i)) return false;
+            }
+            return true;
+        }
+
+        /// <inheritdoc />
+        public override string ToString()
+        {
+            var builder = new StringBuilder();
+
+            builder.Append('(');
+            for (var i = 0; i < Dimensions; i++)
+            {
+                if (i > 0) builder.Append(", ");
+                builder.Append(LowerLeft.ElementAt(i));
+            }
+            builder.Append(')');
+
+            if (!IsPointInternal())
+            {
+                builder.Append(",(");
+                for (var i = 0; i < Dimensions; i++)
+                {
+                    if (i > 0) builder.Append(", ");
+                    builder.Append(UpperRight.ElementAt(i));
+                }
+                builder.Append(')');
+            }
+
+            return builder.ToString();
+        }
+
+        public bool Equals(NpgsqlCube other) => 
+            Dimensions == other.Dimensions && Coordinates.SequenceEqual(other.Coordinates) && Point == other.Point;
+
+        public override bool Equals(object? obj) => obj is NpgsqlCube other && Equals(other);
+
+        public static bool operator ==(NpgsqlCube x, NpgsqlCube y) => x.Equals(y);
+
+        public static bool operator !=(NpgsqlCube x, NpgsqlCube y) => !(x == y);
+        
+        public override int GetHashCode() => 
+            Dimensions.GetHashCode() * Coordinates.GetHashCode() * Point.GetHashCode();
+
+        static List<double> ParseCorner(string coordinates) =>
+            coordinates.Split(',')
+                .Select(v => double.Parse(v, NumberStyles.Any, CultureInfo.InvariantCulture.NumberFormat)).ToList();
+
+        /// <summary>
+        /// Parses a cube in PostgreSQL's text format.
+        /// </summary>
+        /// <param name="s"></param>
+        /// <returns></returns>
+        public static NpgsqlCube Parse(string s)
```

Comment:
We've had recurring conversations whether to have formatting/parsing on Npgsql-provided types... We definitely want to avoid that in cases where there's a lot of complexity (e.g. date/time types), and the regex above doesn't make me super at-ease either - though this may be simple enough to be OK...

If we do this, we may want to accept a CultureInfo to allow users to specify a culture other than InvariantCulture.

Opinions @vonzshik @Brar @NinoFloris ?

Reply:
Yeah, that regex was terrifying to write and test, but it does work. I'm totally open to suggestions on how to improve this. In the PostgreSQL source they seem to have a custom syntax tree implementation for parsing this, I could make a mirror implementation in pure C# to do this instead.

Reply:
> users may want to parse a cube that they have stored in a string somewhere else.

But if it's stored somewhere else, why would it necessarily be in the PG format? If it's coming from PG, Npgsql can be used to read it in binary (which is what this PR does), but I'm not sure which scenario would have users dump string representations of cubes to some text files, and then want to read it back...

> Formatting on the other hand is probably necessary so we can maintain parity with how Postgres displays cubes, this will make debugging easier for users who are used to that display format.

So I'm not sure that maintaining parity with PG to this extent is a goal... Let's not forget that even .NET List does not have a string format (i.e. ToString), and in the debugger users should be able to simply drill into the lists directly. Another point is that if users want a string representation of a cube, they can simply request from PG directly (as a string).

To argue against myself, the most convincing argument for supporting formatting (but not necessarily parsing) is for display purposes, e.g. in a GUI.

I know I'm being pretty "conservative" on this point, but we've had some issues with complicated string formatting/parsing on other types, and it's not clear to what extent users actually need it. Another point is that this is also something we can add later upon request.

Opinions @vonzshik @Brar @NinoFloris ?

Reply:
@roji For now, I have decided to remove the parsing/formatting code following your more conservative approach. We can revisit this later if it's requested.

Reply:
I agree with @roji: since we don't know whether this functionality is actually required, it's better to be conservative for now.